src/NumericalAlgorithms/Interpolation/IrregularInterpolant.hpp

```diff
@@ -0,0 +1,103 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <cstddef>
+
+#include "DataStructures/Matrix.hpp"
+#include "DataStructures/Tensor/TypeAliases.hpp"
+#include "DataStructures/Variables.hpp"
+#include "ErrorHandling/Assert.hpp"
+#include "Utilities/Blas.hpp"
+#include "Utilities/Gsl.hpp"
+
+/// \cond
+class DataVector;
+template <size_t Dim>
+class Index;
+namespace PUP {
+class er;
+}  // namespace PUP
+// IWYU pragma: no_forward_declare Variables
+/// \endcond
+
+/// \ingroup NumericalAlgorithmsGroup
+/// Interpolates a `Variables` onto an arbitrary set of points.
+template <size_t Dim>
+class IrregularInterpolant {
+ public:
+  IrregularInterpolant(
+      const Index<Dim>& mesh,
+      const tnsr::I<DataVector, Dim, Frame::Logical>& points) noexcept;
+  IrregularInterpolant() = default;
+
+  // clang-tidy: no runtime references
+  /// Serialization for Charm++
+  void pup(PUP::er& p) noexcept;  // NOLINT
+
+  //@{
+  /// Performs the interpolation on a `Variables` with grid points corresponding
+  /// to the `Index<Dim>` specified in the constructor.
+  /// The result is a `Variables` whose internal `DataVector` goes over the
+  /// list of points that were specified in the constructor.
+  /// \note for the void function, `result` will be resized to the proper size.
+  template <typename TagsList>
+  void interpolate(gsl::not_null<Variables<TagsList>*> result,
+                   const Variables<TagsList>& u) const noexcept;
+  template <typename TagsList>
+  Variables<TagsList> interpolate(const Variables<TagsList>& u) const noexcept;
+  //@}
+
+ private:
+  // clang-tidy: redundant declaration.
+  template <size_t Dimm>
+  friend bool operator==(const IrregularInterpolant<Dimm>& /*lhs*/,  // NOLINT
+                         const IrregularInterpolant<Dimm>& /*rhs*/) noexcept;
+  Matrix interpolation_matrix_;
+};
+
+template <size_t Dim>
+template <typename TagsList>
+void IrregularInterpolant<Dim>::interpolate(
+    gsl::not_null<Variables<TagsList>*> result,
+    const Variables<TagsList>& u) const noexcept {
+  // For matrix multiplication of Interp . Source = Result:
+  //   matrix Interp is m rows by k columns
+  //   matrix Source is k rows by n columns
+  //   matrix Result is m rows by n columns
+  const size_t m = interpolation_matrix_.rows();
+  const size_t k = interpolation_matrix_.columns();
+  const size_t n = u.number_of_independent_components;
+  ASSERT(k == u.number_of_grid_points(),
+         "Number of grid points in source 'u', "
+             << u.number_of_grid_points()
+             << ",\n disagrees with the size of the mesh, " << k
+             << ", that was passed into the constructor");
+  if (result->number_of_grid_points() != m) {
+    *result = Variables<TagsList>(m, 0.);
+  }
+  dgemm_('n', 'n', m, n, k, 1.0, interpolation_matrix_.data(), m, u.data(), k,
```

Comment:
Why do you do this with lapack instead of blaze (e.g. https://bitbucket.org/blaze-lib/blaze/wiki/Matrix-Matrix%20Multiplication)? I'm not saying this isn't OK; I'd just like to understand.

Reply:
In this case it would probably all work out and be the better choice, but it's a bit situation dependent (dynamically depneding on AMR). LIBXSMM is quite a bit faster than Blaze for small matrices, but gets slower as they increase because at that point using the MKL is preferred. What really we want is to have  a check if the matrix is "small enough" that LIBXSMM will be faster, otherwise fall back to blas directly. Much of the Blaze Matrix-Vector multiplies will be intentionally broken soon because we need to keep track of whether we are dealing with the spectral or nodal representation.